File: python/compresr/integrations/langchain/retriever.py
"""``BaseDocumentCompressor`` for ``ContextualCompressionRetriever``.

Drop-in replacement for ``LLMChainExtractor`` — one Compresr batch call
replaces N LLM extraction calls.

Example::

    from langchain.retrievers import ContextualCompressionRetriever
    from compresr.integrations.langchain import CompresrExtractor

    compressor = CompresrExtractor(api_key=os.environ["COMPRESR_API_KEY"])
    retriever = ContextualCompressionRetriever(
        base_compressor=compressor,
        base_retriever=vectorstore.as_retriever(search_kwargs={"k": 8}),
    )
"""

from __future__ import annotations

import logging
from typing import Any, List, Optional, Sequence

from .._shared import (
    BATCH_LIMIT,
    DEFAULT_MIN_TOKENS,
    DEFAULT_MODEL,
    DEFAULT_POLICY,
    DEFAULT_RATIO,
    ErrorPolicy,
    build_client,
    estimate_tokens,
)

logger = logging.getLogger(__name__)

try:
    from langchain_core.callbacks.manager import Callbacks  # type: ignore[import-not-found]
    from langchain_core.documents import Document  # type: ignore[import-not-found]
except ImportError as exc:  # pragma: no cover
    raise ImportError(
        "CompresrExtractor requires langchain-core. "
        "Install with: pip install compresr[langchain]"
    ) from exc

# `BaseDocumentCompressor` lives in `langchain_core.documents` through
# LangChain 1.x; some 1.x distributions also re-export it from
# `langchain_classic`. Prefer `langchain_core` to avoid pulling the heavy
# `langchain_classic` import chain (transformers/keras) when not needed.
try:
    from langchain_core.documents import (  # type: ignore[import-not-found]
        BaseDocumentCompressor,
    )
except ImportError:  # pragma: no cover
    try:
        from langchain_classic.retrievers.document_compressors.base import (  # type: ignore[import-not-found]
            BaseDocumentCompressor,
        )
    except ImportError as exc:
        raise ImportError(
            "CompresrExtractor requires `BaseDocumentCompressor`. "
            "Install `langchain-core>=0.3` or `langchain-classic`."
        ) from exc


class CompresrExtractor(BaseDocumentCompressor):
    """Compress retrieved ``Document`` content with query-aware compression."""

    api_key: Optional[str] = None
    base_url: Optional[str] = None
    compression_model: str = DEFAULT_MODEL
    target_compression_ratio: float = DEFAULT_RATIO
    min_tokens: int = DEFAULT_MIN_TOKENS
    coarse: Optional[bool] = None
    on_error: ErrorPolicy = DEFAULT_POLICY
    drop_below_min: bool = False
    client: Any = None

    model_config = {"arbitrary_types_allowed": True}

    def _get_client(self) -> Any:
        if self.client is not None:
            return self.client
        self.client = build_client(
            api_key=self.api_key,
            base_url=self.base_url,
            caller="CompresrExtractor",
        )
        return self.client

    def _partition(self, documents: Sequence[Document]) -> tuple[list[int], list[Document]]:
        indices: list[int] = []
        eligible: list[Document] = []
        for i, doc in enumerate(documents):
            if (
                isinstance(doc.page_content, str)
                and estimate_tokens(doc.page_content) >= self.min_tokens
            ):
                indices.append(i)
                eligible.append(doc)
        return indices, eligible

    def _emit(self, doc: Document, new_content: str) -> Document:
        return Document(
            page_content=new_content,
            metadata={**doc.metadata, "compresr": True},
        )

    def _apply_results(
        self,
        out: List[Document],
        documents: Sequence[Document],
        indices: list[int],
        results: list,
        slice_start: int,
        chunk_len: int,
    ) -> None:
        for offset, item in enumerate(results[:chunk_len]):
            global_idx = indices[slice_start + offset]
            new = getattr(item, "compressed_context", None)
            if isinstance(new, str) and new:
                out[global_idx] = self._emit(documents[global_idx], new)

    def compress_documents(
        self,
        documents: Sequence[Document],
        query: str,
        callbacks: Optional[Callbacks] = None,
    ) -> List[Document]:
        if not documents:
            return []

        indices, eligible = self._partition(documents)
        if not eligible:
            return list(documents)

        client = self._get_client()
        out: List[Document] = list(documents)

        for start in range(0, len(eligible), BATCH_LIMIT):
            chunk = eligible[start : start + BATCH_LIMIT]
            try:
                resp = client.compress_batch(
                    contexts=[d.page_content for d in chunk],
                    queries=query,
                    compression_model_name=self.compression_model,
                    target_compression_ratio=self.target_compression_ratio,
                    coarse=self.coarse,
                )
            except Exception as exc:  # noqa: BLE001
                if self.on_error == "raise":
                    raise
                logger.warning(
                    "compresr: batch compress failed (%s); passthrough %d docs.",
                    exc,
                    len(chunk),
                )
                continue

            results = (resp.data.results if resp.data else None) or []
            self._apply_results(out, documents, indices, results, start, len(chunk))

        if self.drop_below_min:
            return [d for d in out if d.page_content and d.page_content.strip()]
        return out

    async def acompress_documents(
        self,
        documents: Sequence[Document],
        query: str,
        callbacks: Optional[Callbacks] = None,
    ) -> List[Document]:
        if not documents:
            return []

        indices, eligible = self._partition(documents)
        if not eligible:
            return list(documents)

        client = self._get_client()
        out: List[Document] = list(documents)

        for start in range(0, len(eligible), BATCH_LIMIT):
            chunk = eligible[start : start + BATCH_LIMIT]
            try:
                resp = await client.compress_batch_async(
                    contexts=[d.page_content for d in chunk],
                    queries=query,
                    compression_model_name=self.compression_model,
                    target_compression_ratio=self.target_compression_ratio,
                    coarse=self.coarse,
                )
            except Exception as exc:  # noqa: BLE001
                if self.on_error == "raise":
                    raise
                logger.warning(
                    "compresr: async batch compress failed (%s); passthrough %d docs.",
                    exc,
                    len(chunk),
                )
                continue

            results = (resp.data.results if resp.data else None) or []
            self._apply_results(out, documents, indices, results, start, len(chunk))

        if self.drop_below_min:
            return [d for d in out if d.page_content and d.page_content.strip()]
        return out

File: typescript/src/integrations/langchain/retriever.ts
/**
 * `BaseDocumentCompressor` implementation for `ContextualCompressionRetriever`.
 *
 * Mirrors Python `CompresrExtractor`. One Compresr batch call
 * replaces N LLM extraction calls.
 */
import { Document } from '@langchain/core/documents';

import type { CompressionClient } from '../../clients/compression.js';
import { getLogger } from '../../logger.js';
import {
  BATCH_LIMIT,
  buildClient,
  DEFAULT_MIN_TOKENS,
  DEFAULT_MODEL,
  DEFAULT_POLICY,
  DEFAULT_RATIO,
  type ErrorPolicy,
  estimateTokens,
} from '../_shared/index.js';

type DocumentType = Document;

export interface CompresrExtractorOptions {
  apiKey?: string;
  client?: CompressionClient;
  baseUrl?: string;
  compressionModel?: string;
  targetCompressionRatio?: number;
  minTokens?: number;
  coarse?: boolean;
  onError?: ErrorPolicy;
  /** Drop documents whose compressed content is empty. */
  dropBelowMin?: boolean;
}

/**
 * Compress retrieved `Document`s with query-aware compression.
 *
 * @example
 * ```ts
 * import { ContextualCompressionRetriever } from '@langchain/classic/retrievers/contextual_compression';
 * import { CompresrExtractor } from '@compresr/sdk/integrations/langchain';
 *
 * const compressor = new CompresrExtractor({
 *   apiKey: process.env.COMPRESR_API_KEY!,
 * });
 * const retriever = new ContextualCompressionRetriever({
 *   baseCompressor: compressor,
 *   baseRetriever: vectorStore.asRetriever({ k: 8 }),
 * });
 * ```
 */
export class CompresrExtractor {
  private readonly compresr: CompressionClient;
  private readonly compressionModel: string;
  private readonly targetCompressionRatio: number;
  private readonly minTokens: number;
  private readonly coarse: boolean | undefined;
  private readonly onError: ErrorPolicy;
  private readonly dropBelowMin: boolean;

  constructor(options: CompresrExtractorOptions) {
    const {
      apiKey,
      client,
      baseUrl,
      compressionModel = DEFAULT_MODEL,
      targetCompressionRatio = DEFAULT_RATIO,
      minTokens = DEFAULT_MIN_TOKENS,
      coarse,
      onError = DEFAULT_POLICY,
      dropBelowMin = false,
    } = options;

    this.compresr =
      client ?? buildClient({ apiKey, baseUrl, caller: 'CompresrExtractor' });
    this.compressionModel = compressionModel;
    this.targetCompressionRatio = targetCompressionRatio;
    this.minTokens = minTokens;
    this.coarse = coarse;
    this.onError = onError;
    this.dropBelowMin = dropBelowMin;
  }

  private partition(
    documents: readonly DocumentType[]
  ): { indices: number[]; eligible: DocumentType[] } {
    const indices: number[] = [];
    const eligible: DocumentType[] = [];
    documents.forEach((doc, i) => {
      const pc = (doc as { pageContent?: unknown }).pageContent;
      if (typeof pc === 'string' && estimateTokens(pc) >= this.minTokens) {
        indices.push(i);
        eligible.push(doc);
      }
    });
    return { indices, eligible };
  }

  private emit(doc: DocumentType, content: string): DocumentType {
    const md = (doc as { metadata?: Record<string, unknown> }).metadata ?? {};
    return new Document({
      pageContent: content,
      metadata: { ...md, compresr: true },
    });
  }

  async compressDocuments(
    documents: readonly DocumentType[],
    query: string
  ): Promise<DocumentType[]> {
    if (documents.length === 0) return [];

    const { indices, eligible } = this.partition(documents);
    if (eligible.length === 0) return [...documents];

    const out: DocumentType[] = [...documents];

    for (let start = 0; start < eligible.length; start += BATCH_LIMIT) {
      const chunk = eligible.slice(start, start + BATCH_LIMIT);
      let results: { compressed_context: string }[];
      try {
        const resp = await this.compresr.compressBatch({
          contexts: chunk.map((d) => (d as { pageContent: string }).pageContent),
          queries: query,
          compressionModelName: this.compressionModel,
          targetCompressionRatio: this.targetCompressionRatio,
          ...(this.coarse !== undefined ? { coarse: this.coarse } : {}),
        });
        results = (resp.data?.results ?? []);
      } catch (exc) {
        if (this.onError === 'raise') throw exc;
        getLogger().warn(
          `batch compress failed (${stringifyError(exc)}); passthrough ${chunk.length} docs.`
        );
        continue;
      }

      results.slice(0, chunk.length).forEach((item, offset) => {
        const globalIdx = indices[start + offset];
        const newText = item?.compressed_context;
        if (typeof newText === 'string' && newText && globalIdx !== undefined) {
          const doc = documents[globalIdx];
          if (doc !== undefined) {
            out[globalIdx] = this.emit(doc, newText);
          }
        }
      });
    }

    if (this.dropBelowMin) {
      return out.filter((d) => {
        const pc = (d as { pageContent?: unknown }).pageContent;
        return typeof pc === 'string' && pc.trim().length > 0;
      });
    }
    return out;
  }
}

function stringifyError(exc: unknown): string {
  return exc instanceof Error ? exc.message : String(exc);
}
